test/benchmark/benchmark_malloc_realloc.cpp

```diff
@@ -0,0 +1,144 @@
+// measurements.
+BENCHMARK_DEFINE_F(PerformanceTestFixture, benchmark_on_malloc)(
+  benchmark::State & state)
+{
+  const size_t malloc_size = state.range(1);
+  if (malloc_size < 1) {
+    state.SkipWithError("Size for allocation is too small for this test");
+  }
+  for (auto _ : state) {
+    void * ptr = std::malloc(malloc_size);
+    if (PERFORMANCE_TEST_FIXTURE_UNLIKELY(nullptr == ptr)) {
+      state.SkipWithError("Malloc failed to malloc");
+    }
+    std::free(ptr);
+    benchmark::DoNotOptimize(ptr);
+    benchmark::ClobberMemory();
+  }
+}
+
+BENCHMARK_DEFINE_F(PerformanceTestFixture, benchmark_on_realloc)(
+  benchmark::State & state)
+{
+  const int malloc_size = state.range(1);
+  if (malloc_size < 1) {
+    state.SkipWithError("Size for allocation is too small for this test");
+  }
+  const size_t realloc_size = state.range(2);
+  if (realloc_size < 1) {
+    state.SkipWithError("Size for reallocation is too small for this test");
+  }
+  for (auto _ : state) {
+    void * ptr = std::malloc(malloc_size);
+    if (PERFORMANCE_TEST_FIXTURE_UNLIKELY(nullptr == ptr)) {
+      state.SkipWithError("Malloc failed to malloc");
+    }
+    void * realloc_ptr = std::realloc(ptr, realloc_size);
+    if (PERFORMANCE_TEST_FIXTURE_UNLIKELY(nullptr == realloc_ptr)) {
+      std::free(ptr);
+      state.SkipWithError("Malloc failed to realloc");
+    }
+    std::free(realloc_ptr);
+    benchmark::DoNotOptimize(realloc_ptr);
+    benchmark::ClobberMemory();
+  }
+}
+
+// Malloc sizes Range from 1 to 2^27 each time multiplying by 16. Each value tested
+// with/without performance metrics
+static void malloc_args(benchmark::internal::Benchmark * b)
+{
+  for (int shift_left = 0; shift_left < 32; shift_left += 4) {
+    for (int enable_tracking = kDisablePerformanceTracking;
+      enable_tracking <= kEnablePerformanceTracking; enable_tracking++)
+    {
+      b->Args({enable_tracking, 1 << shift_left});
+    }
+  }
+}
+
+BENCHMARK_REGISTER_F(PerformanceTestFixture, benchmark_on_malloc)
+->ArgNames({"Enable Performance Tracking", "Malloc Size"})->Apply(malloc_args);
+
+// Three types of realloc tests, one where malloc is smaller than realloc, one where they are
+// the same, and one where malloc is larger than realloc. Realloc size ranges from 1 to 2^27
+// each time multiplying by 32. Each stop is tested with/without performance metrics
+static void realloc_args(benchmark::internal::Benchmark * b)
+{
+  for (int malloc_adjustment = -1; malloc_adjustment <= 1; ++malloc_adjustment) {
+    for (int realloc_shift = 0; realloc_shift < 32; realloc_shift += 8) {
+      for (int enable_tracking = kDisablePerformanceTracking;
+        enable_tracking <= kEnablePerformanceTracking; ++enable_tracking)
+      {
+        const int malloc_shift = realloc_shift + malloc_adjustment;
+        if (malloc_shift < 0) {
+          continue;
+        }
+        b->Args({enable_tracking, 1 << malloc_shift, 1 << realloc_shift});
+      }
+    }
+  }
+}
```

Comment:
Just curious - have you tried using the `->Ranges({ ... })` function to do this instead?

Reply:
I did look at that, but I couldn't find a way to control the spacing when using `Ranges`. Instead benchmark chooses values itself? If I use ranges instead, it generates 22 benchmarks when using a similar range, but they are reasonable spaced. While the number is a bit high, I was most worried about not having control over the benchmark it actually does generate. It seems like we might be at its mercy if we update the vendored version or something?

Reply:
I cleaned up both functions by removing the inner most for loop. It was added originally on the outside, but it can be unrolled on the inside.

Reply:
Makes sense, thanks!

Reply:
Also, it looks like if you have `Ranges` with two lists, it will run all combinations. In the new benchmark for testing without memory tools, I tried `{{1, 1<<27}, {2, 1 << 28}}` and it chose to run so many that the benchmark ultimately timed out.